mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -736,15 +765,27 @@ class GeneralizedLinearRegressionModel private[ml] (
   private lazy val familyAndLink = new FamilyAndLink(familyObj, linkObj)
 
   override protected def predict(features: Vector): Double = {
-    val eta = predictLink(features)
+    if (!isSet(offsetCol) || $(offsetCol).isEmpty) {
+      val eta = BLAS.dot(features, coefficients) + intercept
+      familyAndLink.fitted(eta)
+    } else {
+      throw new SparkException("Must supply offset to predict when offset column is set.")
+    }
+  }
+
+  /**
+   * Calculates the predicted value when offset is set.
+   */
+  protected def predict(features: Vector, offset: Double): Double = {
+    val eta = predictLink(features, offset)
     familyAndLink.fitted(eta)
   }
 
   /**
-   * Calculate the link prediction (linear predictor) of the given instance.
+   * Calculates the link prediction (linear predictor) of the given instance.
    */
-  private def predictLink(features: Vector): Double = {
-    BLAS.dot(features, coefficients) + intercept
+  private def predictLink(features: Vector, offset: Double): Double = {
+    BLAS.dot(features, coefficients) + intercept + offset
```

Comment:
it looks like in the case where family = Gaussian and link = Identity we shouldn't be adding offset?  But then, why would anyone set an offset column... maybe, to make sure this strange case does not happen, we can throw an exception in the validation logic if user accidentally set an offset column in the case that family = Gaussian and link = Identity?